src/uu/lsmem/src/lsmem.rs

```diff
@@ -3,15 +3,543 @@
+            return false;
+        }
+
+        for i in 0..curr_block.nr_zones {
+            if curr_block.zones[i] == ZoneId::ZoneUnknown || curr_block.zones[i] != blk.zones[i] {
+                return false;
+            }
+        }
+    }
+    true
+}
+
+fn memory_block_get_node(path: &PathBuf) -> Result<i32, <i32 as FromStr>::Err> {
+    for entry in fs::read_dir(path).unwrap() {
+        let entry = entry.unwrap();
+        let path = entry.path();
+        let filename = path.to_str().unwrap().split('/').last().unwrap();
+        if path.is_dir() && filename.starts_with("node") {
+            return filename["node".len()..].parse();
+        }
+    }
+    Ok(-1)
+}
+
+fn memory_block_read_attrs(lsmem: &Lsmem, path: &PathBuf) -> MemoryBlock {
+    let mut blk = MemoryBlock::new();
+    blk.count = 1;
+    blk.state = MemoryState::Unknown;
+    let filename = path.to_str().unwrap().split('/').last().unwrap();
+    blk.index = filename[NAME_MEMORY.len()..].parse().unwrap();
+
+    let mut removable_path = path.clone();
+    removable_path.push(PATH_SUB_REMOVABLE);
+    blk.removable = read_file_content::<i32>(&removable_path).is_ok();
+
+    let mut state_path = path.clone();
+    state_path.push(PATH_SUB_STATE);
+    if let Ok(state_raw) = read_file_content::<String>(&state_path) {
+        blk.state = MemoryState::from_str(&state_raw).unwrap();
+    }
+
+    if lsmem.have_nodes {
+        blk.node = memory_block_get_node(path).unwrap();
+    }
+
+    blk.nr_zones = 0;
+    if lsmem.have_zones {
+        if let Ok(raw_content) = read_file_content::<String>(Path::new(PATH_VALID_ZONES)) {
+            let zone_toks = raw_content.split(' ').collect::<Vec<&str>>();
+            for (i, zone_tok) in zone_toks
+                .iter()
+                .enumerate()
+                .take(std::cmp::min(zone_toks.len(), ZoneId::MaxNrZones as usize))
+            {
+                blk.zones[i] = serde_json::from_str(zone_tok).unwrap();
+                blk.nr_zones += 1;
+            }
+        }
+    }
+    blk
+}
+
+fn set_summary(lsmem: &Lsmem) -> Vec<TableRow> {
+    let mut table = Vec::<TableRow>::new();
+
+    for i in 0..lsmem.nblocks {
+        let mut row = TableRow::new();
+
+        let blk = lsmem.blocks[i].borrow();
+
+        // Range
+        let start = blk.index * lsmem.block_size;
+        let size = blk.count * lsmem.block_size;
+        row.range = format!("0x{:016x}-0x{:016x}", start, start + size - 1);
+
+        // Size
+        row.size = if lsmem.bytes {
+            format!("{}", blk.count * lsmem.block_size)
+        } else {
+            utils::size_to_human_string(blk.count * lsmem.block_size)
+        };
+
+        // State
+        row.state = match blk.state {
+            MemoryState::Online => MemoryState::Online.to_string(),
+            MemoryState::Offline => MemoryState::Offline.to_string(),
+            MemoryState::GoingOffline => MemoryState::GoingOffline.to_string(),
+            _ => "?".to_string(),
```

Comment:
As there are only four states, I would list the fourth state, too:
```suggestion
            MemoryState::Unknown => "?".to_string(),
```

Reply:
fixed